### PR TITLE
feat: optimize vector_mix with SSE4.1

### LINKED ISSUE #4: Optimize vector_mix
While `vector_mix` is optimized for most cases with SSE2, we can do better with SSE4.1: `blendps` should be used where we can and the `insertps` and if all else fails, we need a general optimized implementation (2 shuffles?).

Similarly, a NEON implementation should be provided.

### PR DESCRIPTION
**Pull request by operasfantom**
partially fixes #4

### CODE REVIEW

**Review by nfrechette — APPROVED** [review]
Great stuff!
Thank you for the contribution and for keeping the code consistent with the rest. It is much appreciated :)
The static_condition is a bit awkward but sadly necessary for some conditions like this under C++11. It boils down to `if constexpr` but is much more verbose.

**Comment by operasfantom** [comment]
Thank you for the appreciation and thorough review.

As a follow-up, the same optimization can be applied to the alternative version of `vector_mix` operating with double vectors. However, I can't find a way to pass `vector4d` to the corresponding AVX intrinsics. I would expect custom or built-in reinterpret casts, but it seems they are not. Do you have any thoughts on it?

File: includes/rtm/vector4f.h [hunk 1]
```diff
@@ -2139,6 +2139,56 @@ namespace rtm
...
+        // Each component comes from the respective position of input 0 or input 1
+        if (rtm_impl::static_condition<(comp0 == mix4::a || comp0 == mix4::x) && (comp1 == mix4::b || comp1 == mix4::y) &&
+                                       (comp2 == mix4::c || comp2 == mix4::z) && (comp3 == mix4::d || comp3 == mix4::w)>::test()) {
+            constexpr int mask = (comp0 == mix4::a ? 1 : 0) | (comp1 == mix4::b ? 2 : 0) |
+                                 (comp2 == mix4::c ? 4 : 0) | (comp3 == mix4::d ? 8 : 0);
+            return _mm_blend_ps(input0, input1, mask);
+        }
+
+        // First component comes from input 1, others come from the respective positions of input 0
+        if (rtm_impl::static_condition<rtm_impl::is_mix_abcd(comp0) && comp1 == mix4::y && comp2 == mix4::z && comp3 == mix4::w>::test()) {
```
[the remark this answers is not in the document]
Comment (operasfantom) [reply includes/rtm/vector4f.h]: Omitted them for brevity

File: includes/rtm/vector4f.h [hunk 2]
```diff
@@ -2139,6 +2139,56 @@ namespace rtm
 	template<mix4 comp0, mix4 comp1, mix4 comp2, mix4 comp3>
 	RTM_DISABLE_SECURITY_COOKIE_CHECK RTM_FORCE_INLINE vector4f RTM_SIMD_CALL vector_mix(vector4f_arg0 input0, vector4f_arg1 input1) RTM_NO_EXCEPT
 	{
+#if defined(RTM_SSE4_INTRINSICS)
+        // Each component comes from the respective position of input 0 or input 1
+        if (rtm_impl::static_condition<(comp0 == mix4::a || comp0 == mix4::x) && (comp1 == mix4::b || comp1 == mix4::y) &&
+                                       (comp2 == mix4::c || comp2 == mix4::z) && (comp3 == mix4::d || comp3 == mix4::w)>::test()) {
```
[the remark this answers is not in the document]
Comment (operasfantom) [reply includes/rtm/vector4f.h]: Moved braces on their own line

**Comment by nfrechette** [comment]
`vector4d` contains 2x XMM values as opposed to 1x YMM value when AVX is enabled. If you look at the code, you should see how XY and ZW are treated in pairs. As such, their treatment is a bit different.

The reason for this is to maintain decent performance on older hardware. The first few AMD CPUs that supported AVX did not do so natively: AVX instructions would be broken down into 2x 128bit uops, working the upper/lower 128bit portions in parallel. For most code, this is slower than working with 2x 128bit explicitly due to instruction dependencies being more relaxed. Even though code size increases, throughput is improved. Lower end Intel hardware also had pretty poor 256bit performance (e.g. Haswell in older MacBook Pros). These hardware platforms are pretty ancient now in 2022. MSVC also struggled with AVX for years. Even VS2019 has poor codegen when AVX is enabled even if only 128bit XMM types are used (it tries to be clever, and often fails). A case could be made that these are problems of the past and we could look into changing this so that vector4d & friends use native 256bit values instead of 128bit pairs but doing so would likely require a new define value to control which of the two to use.

In particular, the XB1 and PS4 CPUs are older AMD CPUs that poorly handle AVX and those platforms are still actively developed for.

The double types haven't been optimized as much as I haven't had a real need for them yet. I designed the code so that migrating from floats to doubles should be straight forward and I do my best to maintain feature parity.

Overall, I find AVX slower than SSE2 with VS2019 on my Zen2 CPU in my [animation compression library](https://github.com/nfrechette/acl) but I haven't tried with newer CPUs nor with VS2022.

**Comment by operasfantom** [comment]
Awesome! I am pleased to learn the full story about the reason for that design.

Before that, I dealt with other SIMD functions which are mapped to a sequence of instructions. There is, for example, `_mm_loadr_ps` is mapped to `_mm_load_ps` and `_mm_shuffle_ps`. So, yes, AVX instructions are another example of how throughput could be improved by a careful library design. And I tend to agree that `vector4d` && friends can continue to exist unless those are a performance bottleneck or interoperability block. But it is not my call, of course :)

Btw, I also don't have the newest VS2022 to test the performance and may be biased when running on my Ryzen 7 5700U.

**Comment by nfrechette** [comment]
Great stuff, thanks!
If you like this kind of stuff, I highly recommend these resources which I think you might enjoy:
https://travisdowns.github.io/blog/2019/06/11/speed-limits.html
https://www.agner.org/optimize/#manuals

**Pull request merged**